Content.Server/DeviceLinking/Systems/PressurePlateSystem.cs

```diff
@@ -0,0 +1,98 @@
+        SubscribeLocalEvent<PressurePlateComponent, ComponentInit>(OnInit);
+        //SubscribeLocalEvent<PressurePlateComponent, EndCollideEvent>(OnEndCollide);
+        //SubscribeLocalEvent<PressurePlateComponent, StartCollideEvent>(OnStartCollide);
+        SubscribeLocalEvent<PressurePlateComponent, ItemPlacedEvent>(OnItemPlaced);
+        SubscribeLocalEvent<PressurePlateComponent, ItemRemovedEvent>(OnItemRemoved);
+    }
+
+    private void OnItemRemoved(EntityUid uid, PressurePlateComponent component, ref ItemRemovedEvent args)
+    {
+        UpdateState(uid, component);
+    }
+
+    private void OnItemPlaced(EntityUid uid, PressurePlateComponent component, ref ItemPlacedEvent args)
+    {
+        UpdateState(uid, component);
+    }
+
+    private void OnInit(EntityUid uid, PressurePlateComponent component, ComponentInit args)
+    {
+        _signalSystem.EnsureSourcePorts(uid, component.PressedPort, component.ReleasedPort);
+        _appearance.SetData(uid, PressurePlateVisuals.Pressed, false);
+    }
+    //private void OnStartCollide(EntityUid uid, PressurePlateComponent component, ref StartCollideEvent args)
+    //{
+    //    UpdateState(uid, component);
+    //}
+    //
+    //private void OnEndCollide(EntityUid uid, PressurePlateComponent component, ref EndCollideEvent args)
+    //{
+    //    UpdateState(uid, component);
+    //}
+
+    private void UpdateState(EntityUid uid, PressurePlateComponent component)
+    {
+        if (!TryComp<ItemPlacerComponent>(uid, out var itemPlacer))
+            return;
+
+
+        var totalMass = 0f;
+        foreach (var ent in itemPlacer.PlacedEntities)
+        {
+            totalMass += GetEntWeightRecursive(ent);
+        }
+
+        var pressed = totalMass >= component.WeightRequired;
+        if (pressed == component.IsPressed)
+            return;
+
+        component.IsPressed = pressed;
+        _signalSystem.SendSignal(uid, component.StatusPort, pressed);
+        _signalSystem.InvokePort(uid, pressed ? component.PressedPort : component.ReleasedPort);
+
+        _appearance.SetData(uid, PressurePlateVisuals.Pressed, pressed);
+        _audio.PlayPvs(pressed ? component.PressedSound : component.ReleasedSound, uid);
+    }
+
+    /// <summary>
+    /// Recursively calculates the weight of the object, and all its contents, and the contents and its contents...
+    /// </summary>
+    private float GetEntWeightRecursive(EntityUid uid)
+    {
+        var totalMass = 0f;
+        if (Deleted(uid)) return 0f;
+        if (TryComp<PhysicsComponent>(uid, out var physics))
+        {
+            totalMass += physics.Mass;
+        }
+        if (TryComp<EntityStorageComponent>(uid, out var entityStorage))
+        {
+            var storage = entityStorage.Contents;
+            foreach (var ent in storage.ContainedEntities)
+            {
+                totalMass += GetEntWeightRecursive(ent);
+            }
+        }
+        return totalMass;
+    }
```

Comment:
this should be in a separate system from pressure plates imo. also why check only entity storage? should check all containers i feel like, i think there is a method for getting recursive containers on containersystem

Reply:
checking bags and stuff for mice seems too expensive no? considering it will be run every time something with physics (everything) steps on or off the plate

Reply:
I would suggest putting this method in the physics system to be able to directly call the method for an object with physics (any object)

Reply:
Ok, I have put this function in a separate system and also added the weight calculation of ContainerManagerComponent, so inventories are now taken into account too. I'm not sure how to properly design this system and where to place it, so it's in the physics. Correct me in case of anything.


https://github.com/space-wizards/space-station-14/assets/96445749/740a7890-71a8-49df-8a35-eceeca153dfc